controllers/app/controllers/actions_controller.go

```diff
@@ -17,46 +17,139 @@ limitations under the License.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	r.Logger.Info(fmt.Sprintf("action.Name: %v", action.Name))
+
+	if action.Status.Status == "" {
+		action.Status.Status = appv1.Processing
+		r.recorder.Eventf(action, corev1.EventTypeNormal, "ActionProcessing", "Action %s start processing", action.Name)
+		if err := r.Status().Update(ctx, action); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
 
-	// TODO(user): your logic here
+	image := &imagehubv1.Image{}
+	if err := r.Get(ctx, client.ObjectKey{Namespace: DefaultNameSpace, Name: action.Spec.AppName}, image); err != nil {
+		r.recorder.Eventf(image, core.EventTypeNormal, "ImageGetFailed", "Infra %s status is pending", image.Name)
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	r.Logger.Info(fmt.Sprintf("image.Name: %v", image.Name))
+	switch image.Spec.DetailInfo.Actions[image.Name].ActionType {
+	case appv1.KubectlAction:
+		r.actionEngine = NewKubectlEngine(ctx, r.Client, action, image)
+	}
 
+	if err := r.actionEngine.Parse(); err != nil {
+		r.recorder.Eventf(action, core.EventTypeNormal, "ActionParseFailed", "Action %s status is Failed", action.Name)
+		action.Status.Status = appv1.Failed
+		if suberr := r.Status().Update(ctx, action); suberr != nil {
+			return ctrl.Result{}, fmt.Errorf("parse err happened err: %v ; %v", err, suberr)
+		}
+		return ctrl.Result{}, fmt.Errorf("parse err happened err: %v ", err)
+	}
+
+	if err := r.actionEngine.Apply(); err != nil {
+		r.recorder.Eventf(action, core.EventTypeNormal, "ActionApplyFailed", "Action %s status is Failed", action.Name)
+		action.Status.Status = appv1.Failed
+		if suberr := r.Status().Update(ctx, action); suberr != nil {
+			return ctrl.Result{}, fmt.Errorf("apply err happened err: %v ; %v", err, suberr)
+		}
+		return ctrl.Result{}, fmt.Errorf("apply err happened err: %v ", err)
+	}
+
+	r.recorder.Eventf(action, core.EventTypeNormal, "ActionApplySuccess", "Action %s status is applying", action.Name)
+	action.Status.Status = appv1.Success
+	if err := r.Status().Update(ctx, action); err != nil {
+		return ctrl.Result{}, err
+	}
 	return ctrl.Result{}, nil
 }
 
-// SetupWithManager sets up the controller with the Manager.
 func (r *ActionsReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	r.recorder = mgr.GetEventRecorderFor("sealos-action-controller")
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&appv1.Actions{}).
 		Complete(r)
 }
+
+func (ctlEngine *KubectlEngine) Parse() error {
+	actionName := ctlEngine.ActionReq.Spec.ActionName
+	tpl := ctlEngine.imageInfo.Spec.DetailInfo.Actions[actionName].Template
+
+	args := ctlEngine.ActionReq.Spec.Args
+	t, err := template.New("action.yaml").Parse(tpl)
+	if err != nil {
+		return fmt.Errorf("action template parse failed")
+	}
+	var byt bytes.Buffer
+	e := t.Execute(&byt, args)
+	if e != nil {
+		return fmt.Errorf("action template execute failed")
+	}
+	ctlEngine.parseResult = byt.Bytes()
+	return nil
+}
+
+func (ctlEngine *KubectlEngine) Apply() error {
```

Comment:
If actions are executed with administrator privileges, there will be security issues